implementations/rust/ockam/ockam_api/src/session.rs

```diff
@@ -0,0 +1,194 @@
+    pings: JoinSet<(Key, Result<(), Error>)>,
+    replacements: JoinSet<(Key, Result<Address, Error>)>,
+}
+
+#[derive(Debug, Copy, Clone, Encode, Decode)]
+#[rustfmt::skip]
+pub struct Message {
+    #[n(0)] key: Key,
+    #[n(1)] ping: Ping,
+}
+
+impl Medic {
+    pub fn new() -> Self {
+        Self {
+            delay: DELAY,
+            sessions: Arc::new(Mutex::new(Sessions::new())),
+            pings: JoinSet::new(),
+            replacements: JoinSet::new(),
+        }
+    }
+
+    pub fn sessions(&self) -> Arc<Mutex<Sessions>> {
+        self.sessions.clone()
+    }
+
+    pub async fn start(self, ctx: Context) -> Result<(), Error> {
+        let ctx = ctx.new_detached(Address::random_local()).await?;
+        let (tx, rx) = mpsc::channel(32);
+        ctx.start_worker(Collector::address(), Collector(tx))
+            .await?;
+        self.go(ctx, rx).await
+    }
+
+    async fn go(mut self, ctx: Context, mut rx: mpsc::Receiver<Message>) -> ! {
+        let ctx = Arc::new(ctx);
+        loop {
+            log::debug!("check sessions");
+            {
+                let mut sessions = self.sessions.lock().unwrap();
+                for (&key, session) in sessions.iter_mut() {
+                    if session.pings().len() < MAX_FAILURES {
+                        let m = Message::new(session.key());
+                        session.add_ping(m.ping);
+                        log::debug!(%key, ping = %m.ping, "send ping");
+                        let l = {
+                            let v = Encodable::encode(&m).expect("message can be encoded");
+                            let r = route![session.address().clone(), DefaultAddress::ECHO_SERVICE];
+                            let t = TransportMessage::v1(r, Collector::address(), v);
+                            LocalMessage::new(t, Vec::new())
+                        };
+                        let sender = ctx.clone();
+                        self.pings
+                            .spawn(async move { (key, sender.forward(l).await) });
+                        continue;
+                    }
+                    match session.status() {
+                        Status::Up => {
+                            log::warn!(%key, "session unresponsive");
+                            let f = session.replacement(session.address().clone());
+                            session.set_status(Status::Down);
+                            log::info!(%key, "replacing session");
+                            self.replacements.spawn(async move { (key, f.await) });
+                        }
+                        Status::Down => {
+                            log::warn!(%key, "session is down");
+                        }
+                    }
+                }
+            }
+
+            let _ = timeout(self.delay, self.get_results(&mut rx)).await;
+        }
+    }
+
+    async fn get_results(&mut self, rx: &mut mpsc::Receiver<Message>) {
+        loop {
+            tokio::select! {
```

Comment:
Could you please elaborate on how this works? Is that intended that we try to run pings and session reestablishments and receive pongs concurrently? Shouldn't we do that in parallel? Also, `tokio::select` is meant to cancel all futures except the one that finished. So, in a loop, we start polling all of the futures I mentioned above one-by-one (allowing them to make some progress), wait till one of them finishes, abort everything else (and remove all the progress that they may achieved, also not cleaning up any state they may have changed so far), then start all of them again on next loop iteration. Am I wrong about how this works?

Reply:
Operationally there are ping tasks and replacement tasks that execute in the context of tokio `JoinSet`s. Their progress is driven by the tokio executor. The `tokio::select!` call only evaluates cancellation-safe futures which eventually yield a result from those tasks (or a message in case of `mpsc::Receiver::recv`). No progress is removed and no state needs to be cleaned up when the loosing futures care cancelled.

For details see:
- https://docs.rs/tokio/latest/tokio/macro.select.html
- https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html

Reply:
@twittner oh, I guess `JoinSet` is the owner of the futures in this case, so futures are not cancelled in `tokio::select`. Tnx, for the clarification. Do you know if futures are able to progress in the background, or only while `join_next` is called? That wasn't clear for me after reading the doc

Reply:
Also, what are you thoughts about having centralized Medic that tracks all session (as in this PR) vs having dedicated Medic per session? Is there a chance latter approach would look simpler?

Reply:
> Do you know if futures are able to progress in the background, or only while `join_next` is called?

A `JoinSet` spawns the futures as regular tokio tasks and they make progress independently of the `join_next` call.

> Also, what are you thoughts about having centralised Medic that tracks all sessions (as in this PR) vs having dedicated Medic per session?

I think tracking all sessions here is more flexible, for instance when expressing dependencies. Isolated session tracking in actors makes such reasoning comparatively difficult. For example, when a secure channel is shared by multiple sessions, it would be enough to ping it once and restart all sessions depending on it. This is easier to implement when all sessions are treated as data in a single `Medic`.